File: examples/csharp-net48-rpc-client/Handlers/MessageHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MQTTnet;
using MQTTnet.Client;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using DeviceRpcClient.Models;

namespace DeviceRpcClient.Handlers
{
    /// <summary>
    /// Обработчик входящих MQTT-сообщений.
    /// Содержит логику маршрутизации и обработки сообщений по топикам.
    /// </summary>
    public class MessageHandler
    {
        private readonly string _serialNumber;
        private readonly Dictionary<string, PendingRequest> _pendingRequests;
        
        // Callback функции для интеграции с DeviceClient
        private readonly Func<string, Task> _sendAckFunc;
        private readonly Func<string, Task> _sendRequestFunc;
        private readonly Func<string, JObject, string, Task> _sendResultFunc;

        // Топики
        private readonly string _topicRsp;
        private readonly string _topicTsk;
        private readonly string _topicCmt;
        private readonly string _topicEva;

        /// <summary>
        /// Конструктор обработчика сообщений.
        /// </summary>
        /// <param name="serialNumber">Серийный номер устройства (SN).</param>
        /// <param name="pendingRequests">Словарь pending-запросов.</param>
        /// <param name="sendAckFunc">Callback для отправки ACK.</param>
        /// <param name="sendRequestFunc">Callback для отправки REQ.</param>
        /// <param name="sendResultFunc">Callback для отправки RES.</param>
        public MessageHandler(
            string serialNumber,
            Dictionary<string, PendingRequest> pendingRequests,
            Func<string, Task> sendAckFunc,
            Func<string, Task> sendRequestFunc,
            Func<string, JObject, string, Task> sendResultFunc)
        {
            _serialNumber = serialNumber;
            _pendingRequests = pendingRequests;
            _sendAckFunc = sendAckFunc;
            _sendRequestFunc = sendRequestFunc;
            _sendResultFunc = sendResultFunc;

            // Инициализация топиков
            _topicRsp = $"srv/{_serialNumber}/rsp";
            _topicTsk = $"srv/{_serialNumber}/tsk";
            _topicCmt = $"srv/{_serialNumber}/cmt";
            _topicEva = $"srv/{_serialNumber}/eva";
        }

        /// <summary>
        /// Основной метод обработки входящих MQTT-сообщений.
        /// </summary>
        /// <param name="args">Аргументы события получения сообщения.</param>
        public async Task HandleMessageAsync(MqttApplicationMessageReceivedEventArgs args)
        {
            try
            {
                var topic = args.ApplicationMessage.Topic;
                
                // Получение payload
                var payload = args.ApplicationMessage.PayloadSegment.Array != null
                    ? Encoding.UTF8.GetString(
                        args.ApplicationMessage.PayloadSegment.Array,
                        args.ApplicationMessage.PayloadSegment.Offset,
                        args.ApplicationMessage.PayloadSegment.Count)
                    : "";

                // Получение correlationData из User Properties (для совместимости с разными реализациями)
                var correlationData = args.ApplicationMessage.UserProperties?
                    .FirstOrDefault(p => p.Name == "correlationData")?.Value ?? "";

                // Альтернативно: проверяем нативное CorrelationData (MQTT 5)
                if (string.IsNullOrEmpty(correlationData) && args.ApplicationMessage.CorrelationData != null)
                {
                    correlationData = Encoding.UTF8.GetString(args.ApplicationMessage.CorrelationData);
                }

                // Собираем все User Properties в словарь
                var userPropertiesDict = args.ApplicationMessage.UserProperties?
                    .ToDictionary(p => p.Name, p => p.Value) ?? new Dictionary<string, string>();

                Console.WriteLine($"[MSG] Получено сообщение: {topic} | Correlation: {correlationData}");

                // Маршрутизация по топикам
                if (topic == _topicTsk)
                {
                    await HandleTaskAnnouncementAsync(correlationData, userPropertiesDict);
                }
                else if (topic == _topicRsp)
                {
                    await HandleTaskResponseAsync(correlationData, payload, userPropertiesDict);
                }
                else if (topic == _topicCmt)
                {
                    await HandleCommitAsync(correlationData, userPropertiesDict);
                }
                else if (topic == _topicEva)
                {
                    await HandleEventAcknowledgementAsync(correlationData, userPropertiesDict);
                }
                else
                {
                    Console.WriteLine($"[WARN] Неизвестный топик: {topic}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[ERROR] Ошибка при обработке сообщения: {ex.Message}");
                Console.WriteLine($"[ERROR] StackTrace: {ex.StackTrace}");
            }
        }

        /// <summary>
        /// Обработка анонса задачи (TSK) от сервера (Trigger mode).
        /// </summary>
        private async Task HandleTaskAnnouncementAsync(string correlationData, Dictionary<string, string> userProps)
        {
            var methodCode = userProps.ContainsKey("method_code") ? userProps["method_code"] : "unknown";
            Console.WriteLine($"[TSK] Анонс задачи: correlation={correlationData}, method={methodCode}");

            // Опциональное подтверждение получения задачи
            await _sendAckFunc(correlationData);

            // Запрос параметров задачи
            await _sendRequestFunc(correlationData);
        }

        /// <summary>
        /// Обработка ответа сервера с параметрами задачи (RSP).
        /// </summary>
        private async Task HandleTaskResponseAsync(string correlationData, string payload, Dictionary<string, string> userProps)
        {
            var methodCode = userProps.ContainsKey("method_code") ? userProps["method_code"] : "unknown";

            // Удаляем запрос из pending
            if (_pendingRequests.ContainsKey(correlationData))
            {
                _pendingRequests.Remove(correlationData);
            }

            Console.WriteLine($"[RSP] Параметры задачи получены: method={methodCode}, correlation={correlationData}");
            Console.WriteLine($"[RSP] Payload: {payload}");

            // Парсинг и выполнение задачи
            JObject parameters = null;
            if (!string.IsNullOrEmpty(payload))
            {
                try
                {
                    parameters = JObject.Parse(payload);
                }
                catch (JsonException ex)
                {
                    Console.WriteLine($"[WARN] Не удалось распарсить payload: {ex.Message}");
                    parameters = new JObject();
                }
            }
            else
            {
                parameters = new JObject();
            }

            // Выполнение задачи
            var result = await ExecuteTaskAsync(methodCode, parameters);

            // Отправка результата
            await _sendResultFunc(correlationData, result, methodCode);
        }

        /// <summary>
        /// Обработка подтверждения получения результата (CMT) от сервера.
        /// </summary>
        private Task HandleCommitAsync(string correlationData, Dictionary<string, string> userProps)
        {
            var resultId = userProps.ContainsKey("result_id") ? userProps["result_id"] : "unknown";
            Console.WriteLine($"[CMT] Получено подтверждение: result_id={resultId}, correlation={correlationData}");
            Console.WriteLine($"[CMT] RPC-цикл завершён успешно!");
            
            return Task.CompletedTask;
        }

        /// <summary>
        /// Обработка подтверждения события (EVA) от сервера.
        /// </summary>
        private Task HandleEventAcknowledgementAsync(string correlationData, Dictionary<string, string> userProps)
        {
            Console.WriteLine($"[EVA] Подтверждение события получено: correlation={correlationData}");
            return Task.CompletedTask;
        }

        /// <summary>
        /// Выполнение задачи на устройстве.
        /// ⚠️ Здесь реализуйте вашу бизнес-логику обработки команд!
        /// </summary>
        /// <param name="methodCode">Код метода/команды.</param>
        /// <param name="parameters">Параметры задачи (JSON).</param>
        /// <returns>Результат выполнения (JSON).</returns>
        private async Task<JObject> ExecuteTaskAsync(string methodCode, JObject parameters)
        {
            Console.WriteLine($"[EXEC] Выполнение задачи: method={methodCode}");
            Console.WriteLine($"[EXEC] Параметры: {parameters}");

            // ====================================
            // ⚠️ ЗДЕСЬ РЕАЛИЗУЙТЕ ВАШУ ЛОГИКУ!
            // ====================================
            // Примеры методов:
            // - "51" - Открыть что-то
            // - "52" - Закрыть что-то
            // - "69" - Специальная команда
            // - "3001" - Интерактивная команда (требует WebSocket)
            
            // Пример проверки интерактивных методов (3000-3999)
            if (int.TryParse(methodCode, out int code) && code >= 3000 && code <= 3999)
            {
                // Интерактивные команды — нужен активный WS-клиент
                Console.WriteLine($"[EXEC] Интерактивная команда {code} — проверьте наличие WS-клиента");
                // Если WS-клиент не активен, возвращаем ошибку 500 (fail-fast)
                // return new JObject { ["status"] = "error", ["message"] = "No active WebSocket client" };
            }

            // Имитация выполнения работы
            await Task.Delay(1000);

            // Возвращаем успешный результат
            var result = new JObject
            {
                ["status"] = "completed",
                ["data"] = new JObject
                {
                    ["method"] = methodCode,
                    ["executed_at"] = DateTime.UtcNow.ToString("O"),
                    ["success"] = true
                }
            };

            Console.WriteLine($"[EXEC] Задача выполнена: {result}");
            return result;
        }
    }
}

File: examples/c-win-clion-rpc-client/src/message_handler.c
/**
 * @file message_handler.c
 * @brief Маршрутизация и обработка входящих MQTT 5 сообщений.
 *
 * Обработка топиков:
 *   srv/<SN>/tsk — анонс задачи (Trigger mode)
 *   srv/<SN>/rsp — параметры задачи
 *   srv/<SN>/cmt — подтверждение получения результата
 *   srv/<SN>/eva — подтверждение события
 */

#include "message_handler.h"
#include "device_client.h"
#include "config.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* ── Вспомогательные функции ───────────────────────────────── */

/**
 * Извлекает correlation data из свойств MQTT 5 сообщения.
 */
static const char *get_correlation_data(MQTTAsync_message *msg,
                                        char *buf, size_t buf_len)
{
    if (msg->properties.count > 0) {
        MQTTProperty *props = msg->properties.array;
        for (int i = 0; i < msg->properties.count; i++) {
            if (props[i].identifier == MQTTPROPERTY_CODE_CORRELATION_DATA) {
                int len = props[i].value.data.len;
                if (len <= 0 || (size_t)len >= buf_len) {
                    buf[0] = '\0';
                    return buf;
                }
                memcpy(buf, props[i].value.data.data, (size_t)len);
                buf[len] = '\0';
                return buf;
            }
        }
    }
    buf[0] = '\0';
    return buf;
}

/**
 * Извлекает значение User Property по имени.
 */
static const char *get_user_property(MQTTAsync_message *msg,
                                     const char *name,
                                     char *buf, size_t buf_len)
{
    if (msg->properties.count > 0) {
        MQTTProperty *props = msg->properties.array;
        for (int i = 0; i < msg->properties.count; i++) {
            if (props[i].identifier == MQTTPROPERTY_CODE_USER_PROPERTY) {
                if (props[i].value.data.len > 0 &&
                    strncmp(props[i].value.data.data, name,
                            (size_t)props[i].value.data.len) == 0) {
                    int vlen = props[i].value.value.len;
                    if (vlen <= 0 || (size_t)vlen >= buf_len) {
                        buf[0] = '\0';
                        return buf;
                    }
                    memcpy(buf, props[i].value.value.data, (size_t)vlen);
                    buf[vlen] = '\0';
                    return buf;
                }
            }
        }
    }
    buf[0] = '\0';
    return buf;
}

/* ── Обработчики по топикам ────────────────────────────────── */

/**
 * Обработка анонса задачи (TSK) от сервера (Trigger mode).
 */
static void handle_task_announcement(MessageHandlerCtx *ctx,
                                     const char *corr_data,
                                     MQTTAsync_message *msg)
{
    char method_code[32];
    get_user_property(msg, "method_code", method_code, sizeof(method_code));

    printf("[TSK] Task announcement: correlation=%s, method=%s\n",
           corr_data, method_code);

    /* Подтверждение получения (опционально) */
    device_client_send_ack(ctx->client, ctx->serial_number, corr_data);

    /* Запрос параметров задачи */
    device_client_send_request(ctx->client, ctx->serial_number, corr_data);
}

/**
 * Обработка ответа сервера с параметрами задачи (RSP).
 */
static void handle_task_response(MessageHandlerCtx *ctx,
                                 const char *corr_data,
                                 MQTTAsync_message *msg)
{
    char method_code[32];
    get_user_property(msg, "method_code", method_code, sizeof(method_code));

    char payload[MAX_PAYLOAD_LEN] = {0};
    if (msg->payloadlen > 0 && (size_t)msg->payloadlen < sizeof(payload)) {
        memcpy(payload, msg->payload, (size_t)msg->payloadlen);
        payload[msg->payloadlen] = '\0';
    }

    printf("[RSP] Task parameters received: method=%s, correlation=%s\n",
           method_code, corr_data);
    printf("[RSP] Payload: %s\n", payload);

    /*
     * ⚠️ ЗДЕСЬ РЕАЛИЗУЙТЕ ВАШУ БИЗНЕС-ЛОГИКУ!
     * Парсинг payload (JSON) и выполнение задачи.
     */
    printf("[EXEC] Executing task: method=%s\n", method_code);

    /* Имитация результата */
    const char *result = "{\"status\":\"completed\",\"data\":\"success\"}";

    /* Отправка результата */
    device_client_send_result(ctx->client, ctx->serial_number,
                              corr_data, result);
}

/**
 * Обработка подтверждения получения результата (CMT) от сервера.
 */
static void handle_commit(const char *corr_data, MQTTAsync_message *msg)
{
    char result_id[64];
    get_user_property(msg, "result_id", result_id, sizeof(result_id));

    printf("[CMT] Confirmation received: result_id=%s, correlation=%s\n",
           result_id, corr_data);
    printf("[CMT] RPC cycle completed successfully!\n");
}

/**
 * Обработка подтверждения события (EVA) от сервера.
 */
static void handle_event_ack(const char *corr_data)
{
    printf("[EVA] Event confirmation received: correlation=%s\n",
           corr_data);
}

/* ── Публичный API ─────────────────────────────────────────── */

void msg_handler_init(MessageHandlerCtx *ctx, const char *sn,
                      MQTTAsync client)
{
    memset(ctx, 0, sizeof(*ctx));
    snprintf(ctx->serial_number, sizeof(ctx->serial_number), "%s", sn);

    snprintf(ctx->topic_rsp, sizeof(ctx->topic_rsp), "srv/%s/rsp", sn);
    snprintf(ctx->topic_tsk, sizeof(ctx->topic_tsk), "srv/%s/tsk", sn);
    snprintf(ctx->topic_cmt, sizeof(ctx->topic_cmt), "srv/%s/cmt", sn);
    snprintf(ctx->topic_eva, sizeof(ctx->topic_eva), "srv/%s/eva", sn);

    ctx->client = client;
}

void msg_handler_on_message(MessageHandlerCtx *ctx,
                            const char *topic, int topic_len,
                            MQTTAsync_message *message)
{
    char corr_buf[MAX_CORR_DATA_LEN];
    const char *corr_data = get_correlation_data(message,
                                                  corr_buf, sizeof(corr_buf));

    printf("[MSG] Message received: %.*s | Correlation: %s\n",
           topic_len, topic, corr_data);

    if (strncmp(topic, ctx->topic_tsk, (size_t)topic_len) == 0) {
        handle_task_announcement(ctx, corr_data, message);
    } else if (strncmp(topic, ctx->topic_rsp, (size_t)topic_len) == 0) {
        handle_task_response(ctx, corr_data, message);
    } else if (strncmp(topic, ctx->topic_cmt, (size_t)topic_len) == 0) {
        handle_commit(corr_data, message);
    } else if (strncmp(topic, ctx->topic_eva, (size_t)topic_len) == 0) {
        handle_event_ack(corr_data);
    } else {
        printf("[WARN] Unknown topic: %.*s\n", topic_len, topic);
    }
}
